docs/how_it_works.md

```diff
@@ -0,0 +1,142 @@
+
+By itself, an expression doesn't produce a value. It only produces a value once you give it to a
+DataFrame context. What happens to the value(s) it produces depends on which context you hand
+it to:
+
+- `DataFrame.select`: produce a DataFrame with only the result of the given expression
+- `DataFrame.with_columns`: produce a DataFrame like the current one, but also with the result of
+  the given expression
+- `DataFrame.filter`: evaluate the given expression, and if it only returns a single Series, then
+  only keep rows where the result is `True`.
+
+Now let's turn our attention to the implementation.
+
+## Polars implementation
+
+For Polars, Narwhals just "passes everything through". For example consider the following:
+```python
+import polars as pl
+import narwhals as nw
+
+df_pl = pl.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
+df = nw.from_native(df_pl)
+df.select(nw.col('a')+1)
+```
+
+`nw.col('a')` produces a `narwhals.expression.Expr` object, which has a private `_call` method.
+We can call `nw.col('a')._call(pl)`, then the result is actually `pl.col('a')`.
+
+We then let Polars do its thing. Which is nice, but also not particularly interesting.
+How about translating expressions to pandas? Well, it's
+interesting to us, and you're still reading, so maybe it'll be interesting to you too.
+
+## pandas implementation
+
+When we called `nw.col('a')._call(pl)`, we got a Narwhals-compliant Polars namespace.
+The pandas namespace (`pd`) isn't Narwhals-compliant, as the pandas API is very different
+from Polars'. So...Narwhals implements a `PandasNamespace`, which includes the top-level
+Polars functions included in the Narwhals API:
+
+```python
+import narwhals as nw
+from narwhals._pandas_like.namespace import PandasNamespace
+
+pn = PandasNamespace(implementation='pandas')
+nw.col('a')._call(pn)
+```
+The result from the last line above is the same as we'd get from `pn.col('a')`, and it's
+a `narwhals._pandas_like.expression.PandasExpr` object, which we'll call `PandasExpr` for
+short.
+
+`PandasExpr` also have a `_call` method - but this one expects a `PandasDataFrame` as input.
+Recall from above that an expression is a function from a dataframe to a sequence of series.
+The `_call` method gives us that function! Let's see it in action.
+
+Note: the following examples uses `PandasDataFrame` and `PandasSeries`. These are wrappers
+around pandas DataFrame and pandas Series, which are Narwhals-compliant. To get the native
+pandas objects out from inside them, we access `PandasDataFrame._dataframe` and `PandasSeries._series`.
+
+```python
+import narwhals as nw
+from narwhals._pandas_like.namespace import PandasNamespace
+from narwhals._pandas_like.dataframe import PandasDataFrame
+import pandas as pd
+
+pn = PandasNamespace(implementation='pandas')
+
+df_pd = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
+df = PandasDataFrame(df_pd, implementation='pandas')
+expression = pn.col('a') + 1
```

Comment:
Should we use some fancy mkdocs-material to highlight the fact that this such line does not compute anything as is? Maybe [code annotation](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#code-annotations)